File: polling-article-lambda/forex_market_hours.py
from datetime import datetime
from zoneinfo import ZoneInfo


def is_forex_market_open() -> bool:
    now_et = datetime.now(tz=ZoneInfo("America/New_York"))
    day = now_et.weekday()  # Mon=0 ... Sat=5, Sun=6
    minutes_of_day = now_et.hour * 60 + now_et.minute
    five_pm = 17 * 60

    if day == 5:                                 # Saturday
        return False
    if day == 6 and minutes_of_day < five_pm:   # Sunday before 17:00 ET
        return False
    if day == 4 and minutes_of_day >= five_pm:  # Friday at/after 17:00 ET
        return False
    return True

File: ai-trader-frontend/src/lib/forexMarketHours.ts
// Forex market is open Sun 5pm ET → Fri 5pm ET
export function isForexMarketOpen(now = new Date()): boolean {
  const fmt = new Intl.DateTimeFormat("en-US", {
    timeZone: "America/New_York",
    weekday: "short",
    hour: "numeric",
    minute: "numeric",
    hour12: false,
  });

  const parts = Object.fromEntries(fmt.formatToParts(now).map((p) => [p.type, p.value]));
  const day = parts.weekday; // "Sun", "Mon", ..., "Sat"
  const hour = parseInt(parts.hour);
  const minute = parseInt(parts.minute);
  const minutesOfDay = hour * 60 + minute;
  const fivePm = 17 * 60;

  if (day === "Sat") return false;
  if (day === "Sun" && minutesOfDay < fivePm) return false;
  if (day === "Fri" && minutesOfDay >= fivePm) return false;
  return true;
}
